test/jdk/java/util/zip/ZipFile/ReadNonStandardExtraHeadersTest.java

```diff
@@ -0,0 +1,937 @@
+        Files.write(VALID_ANT_ZIP, ANT_ZIP64_UNICODE_EXTRA_ZIP);
+
+    }
+
+    /**
+     * Zip and Jars files to validate we can open
+     */
+    private static Stream<Path> zipFilesToTest() {
+        return Stream.of(VALID_APK, VALID_APACHE_COMPRESS_JAR, VALID_ANT_JAR, VALID_ANT_ZIP);
+    }
+
+    /**
+     * Validate that a Zip file which contains an extra header with a data size
+     * 0f 0 can be opened using ZipFile
+     * @throws IOException if an error occurs
+     */
+    public void zipFilesToTest(Path jar) throws IOException {
+        try (ZipFile zf = new ZipFile(jar.toFile())) {
+            System.out.printf("%s opened%n", jar.toAbsolutePath());
+        } catch (IOException ie) {
+            System.out.printf("%n%n%n$$$$ %s NOT opened%n", jar.toAbsolutePath());
+            throw ie;
+        }
+    }
+
+    /**
+     * Validate that a Zip file which contains an extra header with a data size
+     * 0f 0 can be opened using ZipFS
+     * @throws IOException if an error occurs
+     */
+    public void readZipFSTest(Path jar) throws IOException {
+        try (FileSystem fs = FileSystems.newFileSystem(jar, Map.of())) {
+            System.out.printf("%s opened%n", jar.toAbsolutePath());
+        } catch (IOException ie) {
+            System.out.printf("%n%n%n$$$$ %s NOT opened%n", jar.toAbsolutePath());
+            throw ie;
+        }
+    }
+    /**
+     * Utility method which takes a byte array and converts to byte array
+     * declaration.  For example:
+     * <pre>
+     *     {@code
+     *        var fooJar = Files.readAllBytes(Path.of("foo.jar"));
+     *        var result = createByteArray(fooJar, "FOOBYTES");
+     *        System.out.println(result);
+     *      }
+     * </pre>
+     *
+     * @param bytes A byte array used to create a byte array declaration
+     * @param name  Name to be used in the byte array declaration
+     * @return The formatted byte array declaration
+     */
+    public static String createByteArray(byte[] bytes, String name) {
+        StringBuilder sb = new StringBuilder(bytes.length * 5);
+        Formatter fmt = new Formatter(sb);
+        fmt.format("    public static byte[] %s = {", name);
+        final int linelen = 8;
+        for (int i = 0; i < bytes.length; i++) {
+            if (i % linelen == 0) {
+                fmt.format("%n        ");
+            }
+            fmt.format(" (byte) 0x%x,", bytes[i] & 0xff);
+        }
+        fmt.format("%n    };%n");
+        return sb.toString();
+    }
+
+    public static void main(String[] args) throws Exception {
+        setup();
+        var test = new ReadNonStandardExtraHeadersTest();
+        zipFilesToTest().forEach(path -> {
+            try {
+                test.zipFilesToTest(path);
+                test.readZipFSTest(path);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+}
```

Comment:
I think an upstream has the newline here.

Reply:
I don't see one in https://github.com/openjdk/jdk/commit/13f6450e2e70df4df8bd882def837fbd5bef1524#diff-3f36cef3f473e3b83ba97e882fcc9bbb4b84a36bf56ad118e49584c12e2c3a17 either (and it's not showing up in a diff between the two)

Reply:
17: https://patch-diff.githubusercontent.com/raw/openjdk/jdk17u-dev/pull/1670.patch
jdk: https://github.com/openjdk/jdk/commit/13f6450e2e70df4df8bd882def837fbd5bef1524.patch

in the 17 check the end of the file
```
+    }
+}
\ No newline at end of file
```

Reply:
Ah I see it now. It was introduced by the second commit. The original version I have didn't have this issue.

Reply:
~~~
$ diff -u 8313765.17u{,-02} |tail
++            try {
++                test.zipFilesToTest(path);
++                test.readZipFSTest(path);
++            } catch (IOException e) {
++                throw new RuntimeException(e);
++            }
++        });
++    }
 +}
+\ No newline at end of file
~~~